### PR TITLE
openssh: add UCI support

### PR DESCRIPTION
**Pull request by mhusaam**
Update init script to handle UCI and add a default config

Signed-off-by: Mohd Husaam Mehdi <husaam.mehdi@iopsys.eu>

### CODE REVIEW

**Review by systemcrash — APPROVED** [review]

File: net/openssh/files/sshd.init [hunk 1]
```diff
@@ -20,19 +111,28 @@ start_service() {
 		}
 	done
 	mkdir -m 0700 -p /var/empty
+	mkdir -m 0700 -p /root/.ssh
 
-	local lport=$(awk '/^Port / { print $2; exit }' /etc/ssh/sshd_config)
```
Comment (stokito) [inline net/openssh/files/sshd.init:24]: This is needed to read the Port from sshd_config for mdns.
The Port from UCI may be overridden by the sshd_config.
Not so critical but keep in mind.

File: net/openssh/files/sshd.init [hunk 2]
```diff
@@ -6,8 +6,99 @@ STOP=50
...
+	append_addresses "${ipaddrs}" "${Port}"
+
+	[ "${PasswordAuth}" -eq 0 ] && procd_append_param command -o "PasswordAuthentication no"
+	[ "${RootPasswordAuth}" -eq 1 ] && procd_append_param command -o "PermitRootLogin yes"
+	[ "${RootLogin}" -eq 0 ] && procd_append_param command -o "PermitRootLogin no"
+	[ "${MaxAuthTries}" -gt 0 ] && procd_append_param command -o "MaxAuthTries ${MaxAuthTries}"
+	[ "${mdns}" -ne 0 ] && procd_add_mdns "ssh" "tcp" "${Port}" "daemon=sshd"
+
+	if [ "${IdleTimeout}" -ne 0 ]; then
+		procd_append_param command -o "ClientAliveCountMax 1"
```
Comment (stokito) [inline net/openssh/files/sshd.init:54]: `ClientAliveCountMax 1` may be too small.
By default it's 3 so maybe keep it as is?

https://manpages.debian.org/stretch/openssh-server/sshd_config.5.en.html#ClientAliveCountMax

File: net/openssh/files/sshd.init [hunk 3]
```diff
@@ -6,8 +6,99 @@ STOP=50
...
+{
+	uci_load_validate sshd sshd "$1" "$2" \
+		'PasswordAuth:bool:1' \
+		'RootPasswordAuth:bool:1' \
+		'RootLogin:bool:1' \
+		'Interface:string' \
+		'Port:port:22' \
+		'IdleTimeout:uinteger:0' \
+		'MaxAuthTries:uinteger:3' \
+		'enable:bool:1' \
```
Comment (stokito) [inline net/openssh/files/sshd.init:24]: `enable` -> `enabled`

File: net/openssh/files/sshd.config [hunk 4]
```diff
@@ -0,0 +1,6 @@
+config sshd
+	option enable '1'
```
Comment (stokito) [inline net/openssh/files/sshd.config:2]: `enable` -> `enabled`

**Comment by stokito** [comment]
The PR should be useful given that vendors of modern routers prefer the OpenSSH.

Currently we already have the `/etc/config/dropbear` config that has similar keys:
https://github.com/openwrt/openwrt/blob/main/package/network/services/dropbear/files/dropbear.init#L165

Maybe we can migrate the dropbear config to the sshd. E.g. first check if the dropbear exists and if not read the sshd. And similarly for the OpenSSH but it will install own config and that will cause a conflict which config to use.

Users may change the ssh server e.g. have the Dropbear and then install the OpenSSH. But old settings won't be used now e.g. Port number. Of course a user should know what he doing but still some people may lost an access.

Anyway, that would be extremely useful to support same options that the dropbear supports currently. Well, maybe not all, like the `BannerFile` is not critical at all.

Basically since the Dropbear imitates the sshd. It should be possible to just take its init file and just replace PROG and it must work. If something is not compatible we may patch the Dropbear as I did recently for the ssh-keygen.

**Review by stokito — APPROVED** [review]

File: net/openssh/files/sshd.init (hunk 1 above)
Comment (stokito) [reply net/openssh/files/sshd.init:24]: But generally the problem is that old sshd_config may be overridden by the options from the default config that was installed with a new version of a package.

Comment (mhusaam) [reply net/openssh/files/sshd.init:24]: Okay, thank you.

File: net/openssh/files/sshd.init (hunk 3 above)
Comment (mhusaam) [reply net/openssh/files/sshd.init:24]: It was enable in dropbear so I though I should keep it enable here too.

File: net/openssh/files/sshd.init (hunk 2 above)
Comment (mhusaam) [reply net/openssh/files/sshd.init:54]: This is my logic:

from https://www.man7.org/linux/man-pages/man5/sshd_config.5.html

**ClientAliveCountMax:**
The default value is 3.  If ClientAliveInterval is set to
15, and ClientAliveCountMax is left at the default,
unresponsive SSH clients will be disconnected after
approximately 45 seconds.  Setting a zero
ClientAliveCountMax disables connection termination.

Therefore, to mimic IdleTimeout from dropbear, we set
ClientAliveCountMax to 1 and ClientAliveInterval to
the time we want to wait.

File: net/openssh/files/sshd.config (hunk 4 above)
Comment (mhusaam) [reply net/openssh/files/sshd.config:2]: It was enable in dropbear so I thought I should keep it enable here too.

**Comment by mhusaam** [comment]
> The PR should be useful given that vendors of modern routers prefer the OpenSSH.
> 
> Currently we already have the `/etc/config/dropbear` config that has similar keys: https://github.com/openwrt/openwrt/blob/main/package/network/services/dropbear/files/dropbear.init#L165
> 
> Maybe we can migrate the dropbear config to the sshd. E.g. first check if the dropbear exists and if not read the sshd. And similarly for the OpenSSH but it will install own config and that will cause a conflict which config to use.
> 
> Users may change the ssh server e.g. have the Dropbear and then install the OpenSSH. But old settings won't be used now e.g. Port number. Of course a user should know what he doing but still some people may lost an access.
> 
> Anyway, that would be extremely useful to support same options that the dropbear supports currently. Well, maybe not all, like the `BannerFile` is not critical at all.
> 
> Basically since the Dropbear imitates the sshd. It should be possible to just take its init file and just replace PROG and it must work. If something is not compatible we may patch the Dropbear as I did recently for the ssh-keygen.

Hello, thank you for reviewing this PR. I am sorry for the delayed response.

I have updated the PR as follows:
- I have added support for migrating from dropbear to sshd. So, if someone has an image with dropbear and they upgrade their image to one which has sshd, sshd will look for dropbear config file, and if it is found, it will try to generate sshd config from that, and also try to copy the authorized_keys file.
- I have added support for more dropbear options.
- I have changed the approach from passing command line arguments to sshd daemon, to generating a sshd config file and passing that config file.

Kindly review.

**Comment by stokito** [comment]
Oh, sorry, I didn't noticed that the `IdleTimeout` is a Dropbear's thing https://www.mankier.com/8/dropbear#-I

One thing that I see here that if in a future we'll add the `ClientAliveCountMax` then we may have a conflict and add some priority for them. Also this may confuse that the option from the Dropbear is used for OpenSSH.
What if we simply rename it to `ClientAliveInterval` during migration? Yes, technically this will be not exactly the same and will effectively make the timeout triple bigger but that should not break anything. But code will be much simpler and smaller.
 
The LUCI doesn't show an option for the `IdleTimeout` and it's also not in the default Dropbraer's config. Not sure how many users use it. So maybe it's not that important.

File: net/openssh/files/sshd.init [hunk 5]
```diff
@@ -6,8 +6,351 @@ STOP=50
...
+
+	[ -n "${Interface}" ] && {
+		network_get_ipaddrs_all IpAddrs "${Interface}" || {
+			echo "interface ${Interface} has no physdev or physdev has no suitable ip"
+			return 1
+		}
+	}
+
+	local PidFile="/var/run/${NAME}.${Cfg}.pid"
+	local ConfigFile="/etc/ssh/${NAME}.${Cfg}.conf"
```
Comment (stokito) [inline net/openssh/files/sshd.init:323]: The /etc/ folder is persistent and (just potentially) we may kill NAND flash with writes on each restart.
Maybe instead we can put the generated config into `/var/run/sshd` e.g. into tmpfs.

**Comment by stokito** [comment]
I grepped the `option enable `, `config_get enable ` and `config_get_bool enable ` in packages and we do have a few configs that are using the `enable` instead of `enabled`. So to avoid accidental SSH shell loss after editing of the conf I think it would be better to check **both** `enable` and `enabled`.
In a documentation always use the `enabled`. I also think it would be good to make same for other places that use the `enable`: `atftpd`, `collectd`, `dnsmasq`, `dropbear`, `natmap`, `nut`, `socat`, `sslh`, `openvpn`.
Please let me know what you think about this.

For bool options must be used `config_get_bool` instead of just `config_get`. It will parse `no`/`off` `yes`/`on`.

File: net/openssh/files/sshd.init [hunk 6]
```diff
@@ -6,8 +6,351 @@ STOP=50
...
+
+	procd_set_param respawn
+	procd_close_instance
+}
+
+# for adding trigger
+load_interfaces()
+{
+	config_get Interface "$1" Interface
+	config_get enable "$1" enable 1
```
Comment (stokito) [inline net/openssh/files/sshd.init:347]: config_get_bool enable "$1" enable 1

**Comment by stokito** [comment]
I checked and the Dropbear also has some options that weren't migrated:
* `ForceCommand`
* `rsakeyfile` but it was deprecated
* `keyfile` corresponds to `HostKeyFiles`
* `SSHKeepAlive` corresponds to (again) the `ClientAliveInterval`
* `RecvWindowSize` doesn't have a similar option in sshd

File: net/openssh/files/sshd.init (hunk 5 above)
Comment (mhusaam) [reply net/openssh/files/sshd.init:323]: Thanks for pointing this out, I have updated, kindly review.

File: net/openssh/files/sshd.init (hunk 6 above)
Comment (mhusaam) [reply net/openssh/files/sshd.init:347]: Hello, kindly review:

- added a check for both enable and enabled
- updated to use config_get_bool instead of config_get

**Comment by mhusaam** [comment]
> Oh, sorry, I didn't noticed that the `IdleTimeout` is a Dropbear's thing https://www.mankier.com/8/dropbear#-I
> 
> One thing that I see here that if in a future we'll add the `ClientAliveCountMax` then we may have a conflict and add some priority for them. Also this may confuse that the option from the Dropbear is used for OpenSSH. What if we simply rename it to `ClientAliveInterval` during migration? Yes, technically this will be not exactly the same and will effectively make the timeout triple bigger but that should not break anything. But code will be much simpler and smaller.
> 
> The LUCI doesn't show an option for the `IdleTimeout` and it's also not in the default Dropbraer's config. Not sure how many users use it. So maybe it's not that important.

Hello,

- I think what is necessary is that the meaning of the option is preserved. 
- Also, like you said, IdleTimeout is only for dropbear and ClientAliveCountMax and ClientAliveInterval are only for sshd (OpenSSH).
- So either the user can use dropbear option or the explicit sshd options.
- Accordingly I have **updated the logic to give higher priority to sshd options if they are set explicitly.**

**Comment by mhusaam** [comment]
> I checked and the Dropbear also has some options that weren't migrated:
> 
> * `ForceCommand`
> * `rsakeyfile` but it was deprecated
> * `keyfile` corresponds to `HostKeyFiles`
> * `SSHKeepAlive` corresponds to (again) the `ClientAliveInterval`
> * `RecvWindowSize` doesn't have a similar option in sshd
> 
> This is not critical at all, the PR can be merged and we can back to thos later.

I have added support for the option **keyfile** in init script.
But I did not add it to migration script because uci_validate checks the presence of the file and if the file is not present then init script will not start sshd.
So, if the option is set and somehow the file is removed during upgrade (maybe because it was present in non-persistent storage, I know the chance of that is low) then ssh will not be possible to the device unless defaultreset is done.

As you said, maybe we can come back to this later.

**Comment by mhusaam** [comment]
> When installing the sshd it will copy options from dropbear and remove (!) the dropbear config. But the dropbear will remain started and the port 22 will be busy.
> 
> I think that on migration we should copy options from the dropbear but disable the sshd e.g. `uci set sshd.main.enabled=0`. Then a user can itself go and disable dropbear and enable the sshd and make the `uci commit` and restart the sshd. At least he will be in control and know what he is doing.
> 
> Also it looks like we need a protection from double migration after OpenWrt firmware upgade:
> 
> https://openwrt.org/docs/guide-developer/uci-defaults#ensuring_scripts_don_t_overwrite_custom_settingsimplementing_checks
> 
> The least but not least. What if a user just manualy configured the `/etc/ssh/sshd_config` and then upgreade the `openssh-server` package and now instead of his config the `/etc/config/sshd` is used.
> 
> We may need to add an option `config_path /etc/ssh/sshd_config` that will translated to `Include /etc/ssh/sshd_config`.

Hello,

Sorry for the delayed response.

We need to consider mirgation on upgrade too.

If we disable sshd, and someone upgrades with an image which has sshd (openssh), then sshd will get disabled and there will be no ssh possible to the device.

I have added support for option IncludeConfigFile.

Kindly review.

**Comment by feckert** [comment]
- please rebase the changes
- stash to one commit
- please bump PKG_RELEASE number by one

**Comment by mhusaam** [comment]
Updated, kindly review.

**Comment by GeorgeSapkin** [comment]
@mhusaam still relevant? There are merge conflicts.

**Comment by Rondom** [comment]
I think it will be easier to get this merged if we separate the Uci integration from the migration script. It is debatable, whether one wants the migration script by default of not.

**Comment by mhusaam** [comment]
@GeorgeSapkin conflicts resolved. Kindly review.
@Rondom Removed migration script.

File: net/openssh/Makefile [hunk 7]
```diff
@@ -244,6 +244,8 @@ define Package/openssh-server/install
 	sed -r -i 's,^#(HostKey /etc/ssh/ssh_host_(rsa|ed25519)_key)$$$$,\1,' $(1)/etc/ssh/sshd_config
 	$(INSTALL_DIR) $(1)/etc/init.d
 	$(INSTALL_BIN) ./files/sshd.init $(1)/etc/init.d/sshd
+	$(INSTALL_DIR) $(1)/etc/config
+	$(INSTALL_BIN) ./files/sshd.config $(1)/etc/config/sshd
```
Comment (GeorgeSapkin) [inline net/openssh/Makefile]:
```suggestion
	$(INSTALL_CONF) ./files/sshd.config $(1)/etc/config/sshd
```

File: net/openssh/files/sshd.init [hunk 8]
```diff
@@ -6,8 +6,358 @@ STOP=50
...
+		'Interface:string' \
+		'KexAlgorithms:list(string)' \
+		'MacAlgorithms:list(string)' \
+		'MaxAuthTries:uinteger:6' \
+		'mdns:bool:1' \
+		'PasswordAuth:bool:1' \
+		'Port:port:22' \
+		'RootLogin:bool:1' \
+		'RootPasswordAuth:bool:1' \
+		'UsePAM:bool:0'
```
Comment (GeorgeSapkin) [inline net/openssh/files/sshd.init:34]: Are these supposed to be all possible uci config values? If so, shouldn't they be documented/commented in the default config?

File: net/openssh/files/sshd.init [hunk 9]
```diff
@@ -6,8 +6,358 @@ STOP=50
...
 USE_PROCD=1
 PROG=/usr/sbin/sshd
+NAME=sshd
+TEMPCONF="/tmp/sshd_config.tmp"
+
+. /lib/functions.sh
+. /lib/functions/network.sh
+
+validate_section_sshd()
+{
```
Comment (GeorgeSapkin) [inline net/openssh/files/sshd.init]: Nit-picky, but I'm a bit skeptical about placing opening braces on a new line in shell.

File: net/openssh/files/sshd.init [hunk 10]
```diff
@@ -6,8 +6,358 @@ STOP=50
...
+	local IdleTimeout="${1}"
+
+	# from https://www.man7.org/linux/man-pages/man5/sshd_config.5.html
+	# ClientAliveCountMax:
+	# The default value is 3.  If ClientAliveInterval is set to
+        # 15, and ClientAliveCountMax is left at the default,
+        # unresponsive SSH clients will be disconnected after
+        # approximately 45 seconds.  Setting a zero
+        # ClientAliveCountMax disables connection termination.
```
Comment (GeorgeSapkin) [inline net/openssh/files/sshd.init]: This is indented with spaces, while the rest of the file is tabs.

File: net/openssh/files/sshd.init [hunk 11]
```diff
@@ -6,8 +6,358 @@ STOP=50
...
+	# finalize
+	mkdir -p "$(dirname "${ConfigFile}")"
+	mv "${TEMPCONF}" "${ConfigFile}"
+}
+
+sshd_instance()
+{
+	local IpAddrs
+	local Cfg="$1"
+	local ValidationResult="${2}"
```
Comment (GeorgeSapkin) [inline net/openssh/files/sshd.init]: It looks odd when you mix different variable uses, e.g. `$1` and `${2}`. Pick one style and stick to it, unless it's strictly necessary to use the other, which is not the case here.

File: net/openssh/files/sshd.init [hunk 12]
```diff
@@ -6,8 +6,358 @@ STOP=50
...
+
+	procd_set_param respawn
+	procd_close_instance
+}
+
+# for adding trigger
+load_interfaces()
+{
+	config_get Interface "$1" Interface
+	config_get enable "$1" enable 1
```
Comment (GeorgeSapkin) [inline net/openssh/files/sshd.init]:
```suggestion
	config_get_bool enable "$1" enable 1
```

**Review by GeorgeSapkin — CHANGES_REQUESTED** [review]
On top of all the other small comments, please run a shellcheck on all the scripts, if you haven't already.

File: net/openssh/files/sshd.init (hunk 8 above)
Comment (stokito) [reply net/openssh/files/sshd.init:34]: this is validation

Comment (systemcrash) [reply net/openssh/files/sshd.init:34]: They just look different because they use capitalisation, which is not the common parlance in uci.

File: net/openssh/files/sshd.init (hunk 9 above)
Comment (systemcrash) [reply net/openssh/files/sshd.init]: Sceptical that what might happen? It's valid IIRC.

Comment (GeorgeSapkin) [reply net/openssh/files/sshd.init]: It's valid of course, I'm just not fan of this format.

File: net/openssh/files/sshd.init (hunk 8 above)
Comment (mhusaam) [reply net/openssh/files/sshd.init:34]: Yes, this is validation.

File: net/openssh/Makefile (hunk 7 above)
Comment (mhusaam) [reply net/openssh/Makefile]: Done.

File: net/openssh/files/sshd.init (hunk 9 above)
Comment (mhusaam) [reply net/openssh/files/sshd.init]: Done.

File: net/openssh/files/sshd.init (hunk 10 above)
Comment (mhusaam) [reply net/openssh/files/sshd.init]: Updated. Kindly review.

File: net/openssh/files/sshd.init (hunk 11 above)
Comment (mhusaam) [reply net/openssh/files/sshd.init]: Updated, kindly review.

File: net/openssh/files/sshd.init (hunk 12 above)
Comment (mhusaam) [reply net/openssh/files/sshd.init]: Updated, kindly review.

**Comment by mhusaam** [comment]
Ran shellcheck, it only found benign stuff like "local is not supported".

File: net/openssh/files/sshd.init (hunk 8 above)
Comment (GeorgeSapkin) [reply net/openssh/files/sshd.init:34]: I feel like nobody addressed the part where I mention that all possible config values should be documented in the default config.

**Comment by mhusaam** [comment]
@duttajnu FYI.

File: net/openssh/files/sshd.init (hunk 8 above)
Comment (mhusaam) [reply net/openssh/files/sshd.init:34]: These mostly refer to sshd config manual. I think it is too open ended to specify all possible config values in this case.

Comment (GeorgeSapkin) [reply net/openssh/files/sshd.init:34]: Perhaps add a note of that then?

Comment (mhusaam) [reply net/openssh/files/sshd.init:34]: Added some explanation.

File: net/openssh/files/sshd.config [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+	list AllowUsers 'root'
+	option Port '22'
+	option RootLogin '1'
+	option PasswordAuth '1'
+	option RootPasswordAuth '1'
+#	Please refer to sshd config man page to understand these options, others are explained below
+#	AllowUsers: list of strings, default empty (by list we mean list keyword in uci option, for example see AllowUsers)
+#	BannerFile:file
```
Comment (GeorgeSapkin) [inline net/openssh/files/sshd.config]: Examples usually look like:
```suggestion
	# option BannerFile '/etc/example-banner'
```

File: net/openssh/files/sshd.config [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+# 		ClientAliveCountMax:
+# 		The default value is 3.  If ClientAliveInterval is set to
+# 		15, and ClientAliveCountMax is left at the default,
+# 		unresponsive SSH clients will be disconnected after
+# 		approximately 45 seconds.  Setting a zero
+# 		ClientAliveCountMax disables connection termination.
+#
+# 		Therefore, to mimic IdleTimeout from dropbear, we set
+# 		ClientAliveCountMax to 1 and ClientAliveInterval to
+# 		the time we want to wait
```
Comment (GeorgeSapkin) [inline net/openssh/files/sshd.config]: These are mixing tabs and spaces.

File: net/openssh/files/sshd.config (hunk 13 above)
Comment (mhusaam) [reply net/openssh/files/sshd.config]: Updated, kindly review.

File: net/openssh/files/sshd.config (hunk 14 above)
Comment (mhusaam) [reply net/openssh/files/sshd.config]: Updated, kindly review.

File: net/openssh/files/sshd.init [hunk 15]
```diff
@@ -6,8 +6,337 @@ STOP=50
...
+			if [ -z "${MacList}" ]; then
+				MacList="${MacAlgorithm}"
+			else
+				MacList="${MacAlgorithm},${MacList}"
+			fi
+		done
+
+		append_config "MACs" "${MacList}"
+	fi
+}
```
Comment (systemcrash) [inline net/openssh/files/sshd.init:65]: Something like this to reduce code and indent: 
```sh
set_mac_algorithms() {
	local MacAlgorithms="$1"
	local MacList

	[ -n "$MacAlgorithms" ] || return 0

	for alg in $MacAlgorithms; do
		MacList=${MacList:+$MacList,}$alg
	done

	append_config "MACs" "$MacList"
}
```

Test this out and use this same paradigm for the other functions.

Comment (mhusaam) [reply net/openssh/files/sshd.init:65]: Updated, kindly review.

Comment (systemcrash) [reply net/openssh/files/sshd.init:65]: no change here?

you can apply this to UserList below, CipherList, HostKeyFiles, HostKeyAlgorithmList, KexAlgorithmList, etc also.

Comment (mhusaam) [reply net/openssh/files/sshd.init:65]: I had updated all the functions:
https://github.com/openwrt/packages/compare/4e2261899725490cddfc6fab80b7c0c9ae78e15c..87639356b037508fce7166b107250c7707ee1788

Comment (systemcrash) [reply net/openssh/files/sshd.init:65]: This:

```sh
	for alg in $MacAlgorithms; do
		MacList=${MacList:+$MacList,}$alg
	done
```

is the same as

```sh
	for MacAlgorithm in $MacAlgorithms; do
		if [ -z "${MacList}" ]; then
			MacList="${MacAlgorithm}"
		else
			MacList="${MacAlgorithm},${MacList}"
		fi
	done
```

Comment (systemcrash) [reply net/openssh/files/sshd.init:65]: (except now we store the algorithms in order, and not backwards).

File: net/openssh/files/sshd.init [hunk 16]
```diff
@@ -6,8 +6,337 @@ STOP=50
...
+		# if this is the first iteration, then MacList
+		# should be empty, and we only want to add a
+		# comma if there are multiple values,
+		# this is to avoid extra comma in the end
+		if [ -z "${MacList}" ]; then
+			MacList="${MacAlgorithm}"
+		else
+			MacList="${MacAlgorithm},${MacList}"
+		fi
+	done
```
Comment (GeorgeSapkin) [inline net/openssh/files/sshd.init:62]: IDK, if the order matters, but:
```suggestion
	MacList="${MacAlgorithms// /,}
```
Not sure what the original separator is though.

File: net/openssh/files/sshd.init [hunk 17]
```diff
@@ -6,8 +6,337 @@ STOP=50
...
+
+set_allow_users() {
+	local AllowUsers="${1}"
+	local UserList=""
+
+	[ -n "${AllowUsers}" ] || return
+
+	for User in $AllowUsers; do
+		UserList="${User} ${UserList}"
+	done
```
Comment (GeorgeSapkin) [inline net/openssh/files/sshd.init:75]: Isn't this just:
```suggestion
	UserList="${AllowUsers}"
```
or is does the order need to be reversed?

File: net/openssh/files/sshd.init (hunk 15 above)
Comment (GeorgeSapkin) [reply net/openssh/files/sshd.init:65]: I see this has already been brought up.

File: net/openssh/files/sshd.init (hunk 16 above)
Comment (GeorgeSapkin) [reply net/openssh/files/sshd.init:62]: This applies to all the other loops as well.

Comment (GeorgeSapkin) [reply net/openssh/files/sshd.init:62]: ~~Unless, this doesn't work in dash.~~ Tested and works.

Comment (systemcrash) [reply net/openssh/files/sshd.init:62]: This also works I think, IIRC it flags some shellcheck errors because of the shebang (posix sh and //), or because we're in an init script.

Comment (systemcrash) [reply net/openssh/files/sshd.init:62]: It's less readable though. (There will be plenty of users poking around in this script since ssh is a common utility.) So a code comment `# this does x` is good.

Comment (GeorgeSapkin) [reply net/openssh/files/sshd.init:62]: shellcheck needs dialect specified to detect changes correctly via `--shell=SHELLNAME`.

File: net/openssh/files/sshd.init (hunk 17 above)
Comment (systemcrash) [reply net/openssh/files/sshd.init:75]: depends. Kex and stuff order sets preference IIRC.

File: net/openssh/files/sshd.init (hunk 16 above)
Comment (GeorgeSapkin) [reply net/openssh/files/sshd.init:62]: Yeah a comment can be added.

**Comment by GeorgeSapkin** [comment]
I rebased and fixed merge conflicts.

**Comment by pprindeville** [comment]
> I grepped the `option enable `, `config_get enable ` and `config_get_bool enable ` in packages and we do have a few configs that are using the `enable` instead of `enabled`. So to avoid accidental SSH shell loss after editing of the conf I think it would be better to check **both** `enable` and `enabled`. In a documentation always use the `enabled`. I also think it would be good to make same for other places that use the `enable`: `atftpd`, `collectd`, `dnsmasq`, `dropbear`, `natmap`, `nut`, `socat`, `sslh`, `openvpn`. But in documentation warn that the `enabled` works only for new versions. Please let me know what you think about this.
> 
> For bool options must be used `config_get_bool` instead of just `config_get`. It will parse `no`/`off` `yes`/`on`.

It's worse than that.  We also used `disabled` in some packages.  I believe `0` and `1` are also acceptable values.

File: net/openssh/files/sshd.config [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+	option Port '22'
+	option RootLogin '1'
+	option PasswordAuth '1'
+	option RootPasswordAuth '1'
+
+# Please refer to the sshd_config man page for detailed semantics.
+# Below are UCI usage examples and defaults.
+
+# AllowUsers: list of strings (default: empty)
+# list AllowUsers 'root'
```
Comment (pprindeville) [inline net/openssh/files/sshd.config:13]: If these are subordinate to `config sshd` then please indent them.

File: net/openssh/files/sshd.init [hunk 19]
```diff
@@ -27,19 +356,32 @@ start_service() {
 		}
 	done
 	mkdir -m 0700 -p /var/empty
+	mkdir -m 0700 -p /root/.ssh
```
Comment (pprindeville) [inline net/openssh/files/sshd.init:359]: Should this go here or in `base-files`?

**Review by pprindeville — COMMENTED** [review]
Where are we placing the synthetic files?  Into `/etc/ssh/sshd_config.d/*` or somewhere else?
